edb/pgsql/compiler/relctx.py

```diff
@@ -1019,29 +1020,63 @@ def range_for_typeref(
     return rvar
 
 
+def rel_filter_out(
+    lhs: pgast.SelectStmt, rhs: pgast.SelectStmt,
+    path_id: irast.PathId, *,
+    ctx: context.CompilerContextLevel,
+) -> pgast.SelectStmt:
+    """Filter elements out of the LHS that appear on the RHS"""
+
+    # We unfortunately need to wrap up the LHS in another select so that
+    # this works when the LHS is a set op.
+    rvar = rvar_for_rel(lhs, ctx=ctx)
+    qry = pgast.SelectStmt(from_clause=[rvar])
+    # Plumb it all up...
+    pathctx.put_path_value_rvar(qry, path_id, rvar, env=ctx.env)
+    if path_id.is_objtype_path():
+        pathctx.put_path_source_rvar(
+            qry, path_id, rvar, env=ctx.env)
+    pathctx.put_path_bond(qry, path_id)
+
+    # The useful work: grab the identity from the LHS and do a negated
+    # semi join against the RHS.
+    src_ref = pathctx.get_path_identity_var(
+        qry, path_id=path_id, env=ctx.env)
+    pathctx.get_path_identity_output(
+        rhs, path_id=path_id, env=ctx.env)
+    cond_expr = astutils.new_binop(src_ref, rhs, 'NOT IN')
+    qry.where_clause = astutils.extend_binop(
+        qry.where_clause, cond_expr)
+
+    return qry
+
+
 def range_from_queryset(
         set_ops: Sequence[Tuple[str, pgast.SelectStmt]],
         objname: sn.QualName, *,
+        path_id: Optional[irast.PathId]=None,
         ctx: context.CompilerContextLevel) -> pgast.PathRangeVar:
 
     rvar: pgast.PathRangeVar
 
     if len(set_ops) > 1:
         # More than one class table, generate a UNION/EXCEPT clause.
-        qry = pgast.SelectStmt(
-            all=True,
-            larg=set_ops[0][1]
-        )
+        qry = set_ops[0][1]
 
         for op, rarg in set_ops[1:]:
-            qry.op, qry.rarg = op, rarg
-            qry = pgast.SelectStmt(
-                all=True,
-                larg=qry
-            )
+            if op == 'filter':
```

Comment:
Hmm. I'm curious why the `except` op didn't work here?

Reply:
`except` compares the whole row, and we need to compare just the keys. (The whole row won't match, since it's an UPDATE)